wallet-sample-app-practice/main/Asset/daml/Asset.daml

```diff
@@ -0,0 +1,47 @@
+module Asset where 
+
+import DA.Set as Set
+
+data AssetType = AssetType with 
+    issuer: Party 
+    symbol: Text 
+    fungible: Bool -- true means able to split and merge
+    reference: Optional Text 
+  deriving (Eq, Show) 
+
+template Asset with 
+    assetType: AssetType 
+    owner: Party 
+    amount: Decimal 
+    observers: Set Party
+  where 
+    signatory assetType.issuer, owner 
+    observer observers 
+    ensure (if assetType.fungible then amount > 0.0 else amount == 1.0)
+    -- determined by amount so can't use key?
+
+template AssetTransfer with 
+    asset: Asset 
+    receiver: Party 
+  where 
+    -- need issuer authority (using AssetHoldingAccount)
+    signatory (signatory asset) 
+    observer receiver 
+    -- ensure cannot create AssetTransfer that can't be transfered or rolled back?
+    ensure (if asset.assetType.fungible then asset.amount > 0.0 else asset.amount == 1.0)
+
+    choice Cancel_Transfer: ContractId Asset 
+      controller asset.owner 
+      do 
+        create asset 
```

Comment:
if you're trying to use an intermediate contract (locking by archiving) i'd think of refactoring the code to be more usable - create a transfer choice on the Asset itself

i'd also get back to requirements as well and think why a receiver should accept the transfer (think of how bank transfers work)